custom/src/main/java/com/splunk/opentelemetry/servicename/AppServerServiceNameDetector.java

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright Splunk Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.splunk.opentelemetry.servicename;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.stream.Collectors;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+import javax.annotation.Nullable;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.parsers.SAXParser;
+import javax.xml.parsers.SAXParserFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xml.sax.Attributes;
+import org.xml.sax.SAXException;
+import org.xml.sax.helpers.DefaultHandler;
+
+abstract class AppServerServiceNameDetector extends ServiceNameDetector {
+  private static final Logger log = LoggerFactory.getLogger(AppServerServiceNameDetector.class);
+
+  final ResourceLocator locator;
+  final Class<?> serverClass;
+  final boolean supportsEar;
+
+  AppServerServiceNameDetector(
+      ResourceLocator locator, String serverClassName, boolean supportsEar) {
+    this.locator = locator;
+    this.serverClass = locator.findClass(serverClassName);
+    this.supportsEar = supportsEar;
+  }
+
+  /** Use to ignore default applications that are bundled with the app server. */
+  boolean isValidAppName(Path path) {
+    return true;
+  }
+
+  /** Use to ignore default applications that are bundled with the app server. */
+  boolean isValidResult(Path path, @Nullable String result) {
+    return true;
+  }
+
+  /** Path to directory to be scanned for deployments. */
+  abstract Path getDeploymentDir() throws Exception;
+
+  @Override
+  String detect() throws Exception {
+    if (serverClass == null) {
+      return null;
+    }
+
+    Path deploymentDir = getDeploymentDir();
+    if (deploymentDir == null) {
+      return null;
+    }
+
+    if (Files.isDirectory(deploymentDir)) {
+      log.debug("Looking for deployments in '{}'.", deploymentDir);
+      for (Path path : Files.list(deploymentDir).collect(Collectors.toList())) {
```

Comment:
Just noticed it now: `Files.list()` returns a stream that must be closed; can you run it in a try-with-resources block?